keps/0012-framework-extensions.md

```diff
@@ -0,0 +1,319 @@
+
+- Extending Helm Charts, CNAB or operators
+- Managing repos of extensions
+
+## Proposal
+
+### User Stories
+
+- As an Application Operator, I want to be able to add a custom Plan to an existing FrameworkVersion.
+- As an Application Operator, I want to be able to patch existing templates in an existing FrameworkVersion.
+- As an Application Operator, I want to be able to add or update parameters to an existing FrameworkVersion.
+- As an Application Operator, I want to be able to update my base framework easily.
+
+### Extension Spec
+
+To support extending from a base, `framework.yaml` is extended to support the `extends` keyword:
+
+```yaml
+extends:
+  kudo:
+    framework: "mysql"
+    version: "5.7"
+```
+
+What gets inherited? Everything. A framework defined as:
+
+```yaml
+framework: bar
+extends:
+  kudo:
+    framework: "foo"
+    version: "1.0
+```
+
+will have the EXACT same functionality (plans, parameters, tasks, templates, etc) as the base `foo` framework.
+
+#### Refrencing an object from the base
+
+#### Adding a Task
+
+Tasks can be added via the same `tasks` spec in the `Framework` definition. Templates used to define the task can come from the base framework by pre-fixing `base/` to the template name, or from local templates as defined in [KEP-0009](keps/0009-operator-toolkit.md). The follow example shows a new task called `load-data` defined for the extension framework that uses both a template from the base, with a patch object that is defined in the extention framework.
+
+```yaml
+extends:
+  kudo:
+    framework: "mysql"
+    version: "5.7"
+version: "5.7"
+tasks:
+  load-data:
+    resources:
+      - base/init.yaml
+    patches:
+      - load-data.yaml
+```
+
+#### Adding a Plan
+
+Plans can be added to the extension framework and can reference tasks defined in either the base or the exension framework:
+
+```yaml
+plans:
+  load:
+    steps:
+      - name: load
+        tasks:
+          - load-data
+      - name: cleanup
+        tasks:
+          - load-data
+        delete: true
+```
+
+#### Modifying a Task
+
+When a task is defined in an extended framework, it **replaces** the task from the base. The tasks available are dependent on the base type and are described more in detail in their corresponding sub-section.
+
+In order to better extend and adjust tasks from the base, the `from` keyword will be supported. The `from` directive inside of a named task copies over all resources and patches from the base task of the same name. Resources and patches that overlap with base resource and patch names replace resource and patches already defined by the base.
+
+Presuming there was a task in the base framework defined as:
+
+```yaml
+tasks:
+  init:
+    resources:
+      - init.yaml
+```
+
+An extension framework, that was trying to update the `init` task with a patch captured in `templates/init-patch.yaml` could update the task in different, but equivalent ways:
+
+```yaml
+tasks:
+  init:
+    from: base/init
```

Comment:
So `from` is completely equivalent to `resource`, correct? Why do we need it then?